pkg/compactor/partition_compaction_grouper.go

```diff
@@ -2,37 +2,922 @@ package compactor
 
 import (
 	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"sort"
+	"strings"
+	"time"
 
 	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
 	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/model/labels"
 	"github.com/thanos-io/objstore"
+	thanosblock "github.com/thanos-io/thanos/pkg/block"
 	"github.com/thanos-io/thanos/pkg/block/metadata"
 	"github.com/thanos-io/thanos/pkg/compact"
+
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/storage/tsdb"
+)
+
+var (
+	DUMMY_BLOCK_ID = ulid.ULID{}
 )
 
 type PartitionCompactionGrouper struct {
-	ctx    context.Context
-	logger log.Logger
-	bkt    objstore.InstrumentedBucket
+	ctx                      context.Context
+	logger                   log.Logger
+	bkt                      objstore.InstrumentedBucket
+	acceptMalformedIndex     bool
+	enableVerticalCompaction bool
+	blocksMarkedForNoCompact prometheus.Counter
+	hashFunc                 metadata.HashFunc
+	syncerMetrics            *compact.SyncerMetrics
+	compactorMetrics         *compactorMetrics
+	compactorCfg             Config
+	limits                   Limits
+	userID                   string
+	blockFilesConcurrency    int
+	blocksFetchConcurrency   int
+	compactionConcurrency    int
+
+	doRandomPick bool
+
+	ring               ring.ReadRing
+	ringLifecyclerAddr string
+	ringLifecyclerID   string
+
+	noCompBlocksFunc            func() map[ulid.ULID]*metadata.NoCompactMark
+	partitionVisitMarkerTimeout time.Duration
+
+	ingestionReplicationFactor int
 }
 
 func NewPartitionCompactionGrouper(
 	ctx context.Context,
 	logger log.Logger,
 	bkt objstore.InstrumentedBucket,
+	acceptMalformedIndex bool,
+	enableVerticalCompaction bool,
+	blocksMarkedForNoCompact prometheus.Counter,
+	syncerMetrics *compact.SyncerMetrics,
+	compactorMetrics *compactorMetrics,
+	hashFunc metadata.HashFunc,
+	compactorCfg Config,
+	ring ring.ReadRing,
+	ringLifecyclerAddr string,
+	ringLifecyclerID string,
+	limits Limits,
+	userID string,
+	blockFilesConcurrency int,
+	blocksFetchConcurrency int,
+	compactionConcurrency int,
+	doRandomPick bool,
```

Comment:
Should we get rid of these? We always use true. I cant see a use case where we want to make this configurable

Reply:
This is set to false in unit test. So that we can have a controllable environment for testing.